### PR TITLE
[WIP] Add test coverage to FormController

### PR DESCRIPTION
**Pull request by dsrees**
Closes # (No issue attached, I'd be happy to make one)

#### What has been done to verify that this works as intended?
No code has been changed

#### Why is this the best possible solution? Were any other approaches considered?
Tests are important!

#### Are there any risks to merging this code? If so, what are they?
Nope, no code was changed

#### Do we need any specific form for testing your changes? If so, please attach one.
Still no

**What's this PR?**
In a process to better understand how Collect works and how the project is laid out, I started digging through some of the code and decided to find a crucial class and write unit tests around it. `FormController` seemed to fit the bill and there are 0 tests written around the 1100 some lines of code 

Also, looking to get feedback on these tests. Yes, some of them are relatively simple and don't test much behavior, like the methods that get a single value, but I wanted to get some ground work laid out before I start digging into the more complicated functions.

Please leave feedback if you have any and I will continue to add tests to this PR as I have time

### CODE REVIEW

File: collect_app/src/test/java/org/odk/collect/android/logic/FormControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+
+
+    // UUT
+    private FormController formController;
+
+
+    @Before
+    public void setup() throws Exception {
+        MockitoAnnotations.initMocks(this);
```
Comment (heyjamesknight) [inline collect_app/src/test/java/org/odk/collect/android/logic/FormControllerTest.java]: Add `@RunWith(MockitoJUnitRunner.class)` before the class and kill this line. This gives you validated framework usage (discussion [here](https://stackoverflow.com/questions/10806345/runwithmockitojunitrunner-class-vs-mockitoannotations-initmocksthis/10812752#10812752)).

File: collect_app/src/test/java/org/odk/collect/android/logic/FormControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+import java.io.File;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class FormControllerTest {
```
Comment (heyjamesknight) [inline collect_app/src/test/java/org/odk/collect/android/logic/FormControllerTest.java]: Add `@RunWith(MockitoJUnitRunner.class)` before here.

File: collect_app/src/test/java/org/odk/collect/android/logic/FormControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+
+    @Before
+    public void setup() throws Exception {
+        MockitoAnnotations.initMocks(this);
+        when(mockFormEntryController.getModel()).thenReturn(mockFormEntryModel);
+        when(mockFormEntryModel.getFormIndex()).thenReturn(mockFormIndex);
+        when(mockFormEntryModel.getForm()).thenReturn(mockForm);
+
+        this.formController = new FormController(mockMediaFolder,
```
Comment (heyjamesknight) [inline collect_app/src/test/java/org/odk/collect/android/logic/FormControllerTest.java:59]: If you add `@InjectMocks` before the FormController definition above you can kill this line too (not necessary, just a neat trick).

File: collect_app/src/test/java/org/odk/collect/android/logic/FormControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    FormEntryModel mockFormEntryModel;
+    @Mock
+    FormIndex mockFormIndex;
+    @Mock
+    FormDef mockForm;
+
+
+
+    // UUT
+    private FormController formController;
```
Comment (heyjamesknight) [inline collect_app/src/test/java/org/odk/collect/android/logic/FormControllerTest.java]: Adding `@InjectMocks` here will allow the Mockito Runner to create this automatically for you using the previously defined Mocks.

**Review by heyjamesknight — COMMENTED** [review]
So far so good! Thanks for adding more tests, its greatly appreciated.

**Comment by dsrees** [comment]
thanks @jknightco for the feedback. I'll keep plugging away at these.

**Comment by lognaturel** [comment]
Thanks @danielrees18 for taking this on!

@jknightco is going to be leading a session on Android testing in the context of Collect on Nov 8 and I hope you'll be able to join us! https://forum.opendatakit.org/t/odk-testing-workshop-2017-11-08/10205 If you have any ideas for topics we should make sure are discussed, please share on the thread!

**Comment by heyjamesknight** [comment]
@danielrees18 for sure, let me know if I can help at all as you chug along.

**Comment by heyjamesknight** [comment]
@danielrees18 let me know when you want me to peek at these again!

**Comment by dsrees** [comment]
@jknightco thanks. Sorry I haven't been more diligent with this PR. I was able to add a couple more tests the other day which apparently broke the build! I'll work on getting CircleCI to be green again.

**Comment by heyjamesknight** [comment]
No rush, just let me know :) Mainly I don't want to come in and comment every time you commit just in case its a WIP haha.

File: collect_app/src/main/java/org/odk/collect/android/logic/FormController.java [hunk 5]
```diff
@@ -133,7 +133,7 @@ public static synchronized void initializeJavaRosa(IPropertyManager mgr) {
 
     public FormController(File mediaFolder, FormEntryController fec, File instancePath) {
         this.mediaFolder = mediaFolder;
-        formEntryController = fec;
+        this.formEntryController = fec;
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:136]: Hi. Glad to see new tests. Did you add `.this` for consistency? I suggest it’s better without—that higher signal to noise trumps consistency in this case.

**Comment by lognaturel** [comment]
@dsrees is this something that you'd like to keep working on? Do you have any specific feedback you'd like?

**Comment by dsrees** [comment]
@lognaturel sorry I have neglected this PR. I don't think I'll be able to make updates to this anytime in the near future. I'll close it for now and if anything changes, i'll open up a new PR. Thanks!

**Pull request closed without merging**